dipy/align/imaffine.py

```diff
@@ -0,0 +1,1044 @@
+                                    moving.shape, moving_grid2world)
+
+        if self.dim == 2:
+            self.interp_method = vf.interpolate_scalar_2d
+        else:
+            self.interp_method = vf.interpolate_scalar_3d
+
+        if self.sampling_proportion is None:
+            self.update_pdfs = self.update_pdfs_dense
+            self.update_gradient = self.update_gradient_dense
+            self.samples = None
+            self.ns = 0
+            self.transformed = self.affine_map.transform(self.moving)
+            self.transformed = self.transformed.astype(np.float64)
+        else:
+            self.update_pdfs = self.update_pdfs_sparse
+            self.update_gradient = self.update_gradient_sparse
+            self.transformed = None
+            k = int(np.ceil(1.0 / self.sampling_proportion))
+            shape = np.array(static.shape, dtype=np.int32)
+            self.samples = sample_domain_regular(k, shape, static_grid2world)
+            self.samples = np.array(self.samples)
+            self.ns = self.samples.shape[0]
+            # Add a column of ones (homogeneous coordinates)
+            self.samples = np.hstack((self.samples, np.ones(self.ns)[:, None]))
+            # Sample the static image
+            static_p = self.static_world2grid.dot(self.samples.T).T
+            static_p = static_p[..., :self.dim]
+            self.static_vals, inside = self.interp_method(static, static_p)
+            self.static_vals = np.array(self.static_vals, dtype=np.float64)
+            # Sample the moving image
+            sp_to_moving = self.moving_world2grid.dot(P)
+            moving_p = sp_to_moving.dot(self.samples.T).T
+            moving_p = moving_p[..., :self.dim]
+            self.moving_vals, inside = self.interp_method(moving, moving_p)
+        super(MIMetric, self).setup(self.static, self.moving)
+
+    def _update(self, params, update_gradient=True):
+        r""" Updates marginal and joint distributions and the joint gradient
+
+        The distributions are updated according to the static and transformed
+        images. The transformed image is precisely the moving image after
+        transforming it by the transform defined by the `params` parameters.
+
+        The gradient of the joint PDF is computed only if update_gradient
+        is True.
+
+        Parameters
+        ----------
+        params : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+        update_gradient : Boolean, optional
+            if True, the gradient of the joint PDF will also be computed,
+            otherwise, only the marginal and joint PDFs will be computed.
+            The default is True.
+        """
+        # Get the matrix associated with the params parameter vector
+        M = self.transform.param_to_matrix(params)
```

Comment:
So, my idea for avoiding `distance_and_gradient` is something like:

```
def distance(self, params):
    if np.all(params == self._last_params):
        return self._last_distance
    self._last_params = params
    # Calculate, store, return distance

def gradient(self, params):
     # Calculate necessary precursors for gradient calculation
     distance = self.distance(params)
     # Calculate and return gradient
```

Then, when you are requesting the gradient and distance you just do:

```
return obj.distance(), obj.gradient()
```

with no extra cost.  It also has the advantage of splitting the rather large 'update' function.  What do you think?